### PR TITLE
Add ML-KEM768 KATs from BoringSSL

### PR DESCRIPTION
**Pull request by andrewkdinh**
Add KATs for ML-KEM768 under CCLA from https://boringssl.googlesource.com/boringssl/

**Please advise for how I should credit BoringSSL in the test files.**

These KATs test key generation, encapsulation, and decapsulation for the ML-KEM768 provider.

Relevant notes:
- Added functionality to the ML-KEM provider to export/import. These may not be fully implemented yet (see https://github.com/openssl/project/issues/1036)
- Exposed some more low-level ML-KEM API's to allow for deterministic encapsulation/key generation
- Actually run 'mlkem_internal_test' with `make test`
- Fixes openssl/project#1045

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by andrewkdinh** [comment]
Tagging everybody from the [initial PR](https://github.openssl.org/mbaentsch/openssl/pull/5)

@mattcaswell @baentsch @slontis @paulidale @t8m

**Comment by baentsch** [comment]
Thanks for this work @andrewkdinh ! Could you please re-base so the diff is easier to read?

**Comment by t8m** [comment]
> Thanks for this work @andrewkdinh ! Could you please re-base so the diff is easier to read?

It would make sense to move this to openssl/openssl after the initial PR is merged into the feature branch.

**Comment by andrewkdinh** [comment]
@baentsch I've re-based this now

@t8m Ok, I think I'll need to create a separate PR in openssl/openssl once https://github.com/openssl/openssl/pull/25848 is merged

File: providers/implementations/keymgmt/mlkem_kmgmt.c [hunk 1]
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
...
+                                                 &mkey->pubkey))
+            goto err;
+    } else {
+        debug_print("MLKEMKM generate random keys");
+        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
+                                        &mkey->privkey, mkey->mlkem_ctx)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
+                                                 &mkey->privkey)
+            || ossl_mlkem768_public_from_private(&mkey->pubkey,
+                                                 &mkey->privkey))
```
Comment (baentsch) [inline providers/implementations/keymgmt/mlkem_kmgmt.c]: marshal_public_key missing here?

File: providers/implementations/keymgmt/mlkem_kmgmt.c [hunk 2]
```diff
@@ -580,11 +634,12 @@ const OSSL_DISPATCH ossl_mlkem768_keymgmt_functions[] = {
...
     { OSSL_FUNC_KEYMGMT_DUP, (void (*)(void))mlkem_dup },
     /*
-     * TODO(ML-KEM) don't do for now, see https://github.com/openssl/private/issues/698
-     *  { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))mlkem_imexport_types },
-     *  { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))mlkem_imexport_types },
-     *  { OSSL_FUNC_KEYMGMT_IMPORT, (void (*)(void))mlkem_export },
-     *  { OSSL_FUNC_KEYMGMT_EXPORT, (void (*)(void))mlkem_import },
-     */
+     * TODO(ML-KEM): https://github.com/openssl/private/issues/698
+     * Export/import functionality has been partially implemented.
```
Comment (baentsch) [inline providers/implementations/keymgmt/mlkem_kmgmt.c]: Worthwhile stating here which things are missing?

File: test/evp_test.c [hunk 3]
```diff
@@ -2288,30 +2346,75 @@ static int decapsulate(EVP_TEST *t, EVP_PKEY_CTX *ctx, const char *op,
...
 
         ret = encapsulate(t, kdata->ctx, kdata->op, &wrapped, &wrappedlen,
                           &secret, &secretlen);
         if (ret == 0 || t->err != NULL)
             goto err;
-        ret = decapsulate(t, kdata->ctx, kdata->op, wrapped, wrappedlen,
-                          secret, secretlen);
+
+        /* Also attempt to decrypt if we have the private key */
+        if (found_key || kdata->encoded_priv_key != NULL)
```
Comment (baentsch) [inline test/evp_test.c:2405]: Wouldn't it be better for the test to fail if there's no private key available?

File: test/recipes/30-test_evp.t [hunk 4]
```diff
@@ -130,6 +130,9 @@ my @defltfiles = qw(
                      evppkey_kdf_scrypt.txt
                      evppkey_kdf_tls1_prf.txt
                      evppkey_rsa.txt
+                     evppkey_mlkem768_keygen.txt
```
Comment (baentsch) [inline test/recipes/30-test_evp.t]: shouldn't they be done (pushed to defltfiles) only if no_mlkem is not set?

File: providers/implementations/keymgmt/mlkem_kmgmt.c (hunk 1 above)
Comment (andrewkdinh) [reply providers/implementations/keymgmt/mlkem_kmgmt.c]: Line 559 already writes the encoded public key to `mkey->encoded_pubkey`

```c
if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
```

File: test/evp_test.c (hunk 3 above)
Comment (andrewkdinh) [reply test/evp_test.c:2405]: The tests for ML-KEM-768 encap only have the encoded public key (and no private key)

File: providers/implementations/keymgmt/mlkem_kmgmt.c (hunk 2 above)
Comment (andrewkdinh) [reply providers/implementations/keymgmt/mlkem_kmgmt.c]: The encode/decode logic pretty much matches what I saw in other providers. But from https://github.com/openssl/private/issues/698:

> prove interoperability -- and for that we need reference to a standard (please fill in the above if you can: I didn't find anything, so if we want to do this without a standard (?), we need to agree on/how to do interop testing with other implementations).

I haven't done anything like this

File: test/recipes/30-test_evp.t (hunk 4 above)
Comment (andrewkdinh) [reply test/recipes/30-test_evp.t]: Updated: https://github.com/baentsch/openssl/compare/b8e8d90a0a369a69f399885419df3d793006fe3a..3cc382434d4b8f576fc6213e5741b55ffe6ce396

File: providers/implementations/keymgmt/mlkem_kmgmt.c (hunk 2 above)
Comment (baentsch) [reply providers/implementations/keymgmt/mlkem_kmgmt.c]: ACK. Suggest referencing to https://github.com/openssl/openssl/issues/25885 now (contains draft spec references).

File: test/recipes/30-test_evp.t (hunk 4 above)
Comment (baentsch) [reply test/recipes/30-test_evp.t]: Not quite OK: The new variable name is "no_mlkem" (not declared as such). After fixing that, the tests fail for me (both, test_evp and test_evp_extra): Do I need to do some special setup to run them? I used `make TESTS='test_evp_extra test_evp' test`...

File: providers/implementations/keymgmt/mlkem_kmgmt.c [hunk 5]
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
...
+                                                  &mkey->privkey)
+            || !ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->privkey)
+            || !ossl_mlkem768_marshal_public_key(mkey->encoded_pubkey,
+                                                 &mkey->pubkey))
+            goto err;
+    } else {
+        debug_print("MLKEMKM generate random keys");
+        if (!ossl_mlkem768_generate_key(mkey->encoded_pubkey, gctx->seed,
+                                        &mkey->privkey, mkey->mlkem_ctx)
+            || ossl_mlkem768_marshal_private_key(mkey->encoded_privkey,
```
Comment (baentsch) [inline providers/implementations/keymgmt/mlkem_kmgmt.c]: "!" missing -- wrong failure condition and the reason for "evp_extra" test failiing.

File: providers/implementations/keymgmt/mlkem_kmgmt.c [hunk 6]
```diff
@@ -507,13 +537,33 @@ static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
...
+                                                 &mkey->privkey)
+            || ossl_mlkem768_public_from_private(&mkey->pubkey,
```
Comment (baentsch) [inline providers/implementations/keymgmt/mlkem_kmgmt.c]: "!"

**Comment by baentsch** [comment]
@andrewkdinh fyi, with the three code changes as per the above, the tests all work -- for ML-KEM. Failures in SM2 and FFDHE...

File: test/recipes/30-test_evp.t (hunk 4 above)
Comment (andrewkdinh) [reply test/recipes/30-test_evp.t]: Oops, there was a couple incorrect things after fixing a rebase. All tests succeed for me when I run `make test`... can you try again with the latest commit?

**Comment by andrewkdinh** [comment]
@baentsch All tests pass for me with `make test`. But this can also be verified when I open the PR to openssl/openssl when CI runs

**Comment by baentsch** [comment]
> @baentsch All tests pass for me with `make test`. But this can also be verified when I open the PR to openssl/openssl when CI runs

@andrewkdinh : Confirmed for ML-KEM. Thanks! I'd be all for opening this as a public PR now. One heads-up (nit, I hope), though: These 2 tests (unrelated to MLKEM) fail for me (just did clean build using `./config && make -j && make test` on a vanilla x64 ubuntu box):

```

        # INFO:  @ test/testutil/stanza.c:21
        # Reading ../../test/recipes/30-test_evp_data/evppkey_ffdhe.txt
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "RFC7919 DH tests" tests at line 15
../../util/wrap.pl ../../test/evp_test -config ../../test/default-and-legacy.cnf ../../test/recipes/30-test_evp_data/evppkey_ffdhe.txt => 139
not ok 25 - running evp_test -config ../../test/default-and-legacy.cnf evppkey_ffdhe.txt
# ------------------------------------------------------------------------------
        # INFO:  @ test/testutil/stanza.c:21
        # Reading ../../test/recipes/30-test_evp_data/evppkey_sm2.txt
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "SM2 tests" tests at line 19
        # INFO:  @ test/testutil/stanza.c:122
        # Starting "SM2 key generation tests" tests at line 78
../../util/wrap.pl ../../test/evp_test -config ../../test/default-and-legacy.cnf ../../test/recipes/30-test_evp_data/evppkey_sm2.txt => 139
not ok 78 - running evp_test -config ../../test/default-and-legacy.cnf evppkey_sm2.txt
```


@t8m @mattcaswell: What does it take to merge https://github.com/openssl/openssl/pull/25848 such as to allow @andrewkdinh to raise this code as a public PR towards the feature branch?

**Comment by baentsch** [comment]
@andrewkdinh Please redirect this PR towards the mlkem feature branch now: I'm anxious to (first review, of course, and) land this code (and APIs and provider improvements) there to base the next steps on that (ml-kem-1024, probably to ascertain our approach "generalizes" well). Thanks in advance!

**Comment by andrewkdinh** [comment]
@baentsch I need to wait until your PR is fully synced into feature/ml-kem so that the diff is just my changes (usually <24 hours): https://github.com/openssl/openssl/compare/feature/ml-kem...andrewkdinh:openssl:mlkem-kats

I'll also be looking into the failing tests; strange that it didn't fail for me locally

**Comment by mattcaswell** [comment]
> @baentsch I need to wait until your PR is fully synced into feature/ml-kem so that the diff is just my changes (usually <24 hours): https://github.com/openssl/openssl/compare/feature/ml-kem...andrewkdinh:openssl:mlkem-kats

@baentsch's PR has been merged to the feature/ml-kem branch. You should be able to create your PR now.

**Comment by andrewkdinh** [comment]
Opened https://github.com/openssl/openssl/pull/25938 to replace this PR

**Pull request closed without merging**